WIP: ducttape and wd40 driven porting of vert visc loop

JorgeG94 opened the pull request:
Thank you for contributing to MOM6!


TODO:

- [x] - vertvisc_remnant 
- [x] - remove explicit mappings from inside functions
- [x] - port to DC once I have correct results 
- [x] - move mapping of data to around subroutine calls so that subroutines only allocate local data 
- [x] - cleanup of vertvisc everything 
- [x] - main RK loop 

Before completing your submission, remember to check the following.

* Have I submitted to the correct repository?

  .. image:: media/select_repo.png

  - mom-ocean is only for nodes

  - If you are unsure, send this PR to NOAA-GFDL/MOM6.git

* Have all CI tests pass?

  - Reach out to XXX if you need help resolving these issues.

* Review the MOM6 code style guide

  Some common errors:

  - Indentation

  - Correct index conventions

  - Parentheses to ensure bit reproducibility

* Commit history

  - Is my work based on the latest version?

    If not, then you may need to update your branch, preferably using ``git
    rebase``.

  - Are all of my commits meaningful?

marshallward committed:
Vertvisc: First part of loop

Moving gradually

JorgeG94 committed:
add nvtx ranges and module warpper

JorgeG94 committed:
port loop to openmp for now -> will move to do concurrent. Focused on porting to GPU now

JorgeG94 committed:
more gpu

JorgeG94 committed:
port the vertical velocity to GPUs

JorgeG94 committed:
port the following vertical velocity loops to GPUs

JorgeG94 committed:
more loops using omp target teams, validated

JorgeG94 committed:
fix compute sanitizer death

JorgeG94 committed:
meridional velcoity component ported

JorgeG94 committed:
finish porting vertvisc

JorgeG94 committed:
offload another loop

JorgeG94 committed:
revert the last thing

JorgeG94 committed:
add collapse(2) to the expensive loops

JorgeG94 committed:
fix

JorgeG94 committed:
fix validation

JorgeG94 committed:
updates to porting: find coupling coef

JorgeG94 committed:
almost done, just missingt he killer loop

JorgeG94 committed:
lable the block of death

JorgeG94 committed:
duct taped version of all funcs ported to GPUs

JorgeG94 commented on src/parameterizations/vertical/MOM_vert_friction.F90:
```diff
@@ -774,414 +817,689 @@ subroutine vertvisc(u, v, h, forces, visc, dt, OBC, ADp, CDp, G, GV, US, CS, &
+  end do
 
-  if (associated(ADp%du_dt_str)) then
+  if (is_du_dt_str_associated) then
```
I had to do this probably out of desperation at some point due to some errors I was seeing, need to assess if this is truly necessary

edoyango replied:
it shouldn't be necessary outside of target regions, but useful for within target regions since nvfortran 25.3 segfaults if checking an unallocated array (and maybe unassociated pointers?).

marshallward replied:
This is a very common pattern, and it's on the CPU side, so I think we need to assume that the original form works correctly.

JorgeG94 commented on src/parameterizations/vertical/MOM_vert_friction.F90:
```diff
@@ -774,414 +817,689 @@ subroutine vertvisc(u, v, h, forces, visc, dt, OBC, ADp, CDp, G, GV, US, CS, &
   do k=2,nz
-    if (allocated(visc%Ray_u)) then
+    !if (allocated(visc%Ray_u)) then
+    if (is_ray_u_alloc) then 
```
ditto here as above

JorgeG94 commented on src/parameterizations/vertical/MOM_vert_friction.F90:
```diff
@@ -774,414 +817,689 @@ subroutine vertvisc(u, v, h, forces, visc, dt, OBC, ADp, CDp, G, GV, US, CS, &
+    end do
+
+  !  if (associated(ADp%du_dt_str)) then
+    if (is_du_dt_str_associated) then
```
ditto

JorgeG94 commented on src/parameterizations/vertical/MOM_vert_friction.F90:
```diff
@@ -774,414 +817,689 @@ subroutine vertvisc(u, v, h, forces, visc, dt, OBC, ADp, CDp, G, GV, US, CS, &
-      u(I,j,k) = u(I,j,k) + c1(I,j,k+1) * u(I,j,k+1)
-    endif ; enddo ; enddo
+  !$omp target teams loop collapse(2) private(j,i) 
+    do j=G%isc,G%jec
```
need to revert formatting or decide to keep it as is, ask

JorgeG94 commented on src/parameterizations/vertical/MOM_vert_friction.F90:
```diff
@@ -774,414 +817,689 @@ subroutine vertvisc(u, v, h, forces, visc, dt, OBC, ADp, CDp, G, GV, US, CS, &
   ! compute vertical velocity tendency that arises from GL90 viscosity;
   ! follow tridiagonal solve method as above; to avoid corrupting u,
   ! use ADp%du_dt_visc_gl90 as a placeholder for updated u (due to GL90) until last do loop
+  vert_vel_gl90: block
```
get rid of this block, I don't need locality anymore

JorgeG94 commented on src/parameterizations/vertical/MOM_vert_friction.F90:
```diff
@@ -774,414 +817,689 @@ subroutine vertvisc(u, v, h, forces, visc, dt, OBC, ADp, CDp, G, GV, US, CS, &
-        zDS = zDS + h_a ; if (zDS >= Hmix) exit
-      enddo
-    endif ; enddo ; enddo
+  ! TODO JORGE: refactor this one later
```
no idea what I meant here

JorgeG94 commented on src/parameterizations/vertical/MOM_vert_friction.F90:
```diff
@@ -774,414 +817,689 @@ subroutine vertvisc(u, v, h, forces, visc, dt, OBC, ADp, CDp, G, GV, US, CS, &
-    d1(i,J) = b_denom_1 * b1(i,J)
-    v(i,J,1) = b1(i,J) * (CS%h_v(i,J,1) * v(i,J,1) + surface_stress(i,J))
-  endif ; enddo ; enddo
+  !$omp target enter data map(to: CS, CS%a_v, CS%h_v)
```
probably move this up

marshallward replied:
As above, I'd recommmend moving this to `vertvisc_init` or somewhere similar.

*However*, as we've learned, the control structures are very touchy.  Let's confirm that this version works before attempting to move this.

JorgeG94 commented on src/parameterizations/vertical/MOM_vert_friction.F90:
```diff
@@ -774,414 +817,689 @@ subroutine vertvisc(u, v, h, forces, visc, dt, OBC, ADp, CDp, G, GV, US, CS, &
+
   if (CS%id_GLwork > 0) then
-    if (.not.G%symmetric) &
+  print *, "A"
```
delete prints

JorgeG94 commented on src/parameterizations/vertical/MOM_vert_friction.F90:
```diff
@@ -774,414 +817,689 @@ subroutine vertvisc(u, v, h, forces, visc, dt, OBC, ADp, CDp, G, GV, US, CS, &
 
   call vertvisc_limit_vel(u, v, h, ADp, CDp, forces, visc, dt, G, GV, US, CS)
+  !! TODO jorge: this is the last GPU code
+  !$omp target exit data map(from: b1, c1, d1)
```
optimize this later, right now stay ugly

JorgeG94 commented on src/parameterizations/vertical/MOM_vert_friction.F90:
```diff
@@ -1285,10 +1604,16 @@ subroutine vertvisc_remnant(visc, visc_rem_u, visc_rem_v, dt, G, GV, US, CS)
     ! Ray is the Rayleigh-drag velocity [H T-1 ~> m s-1 or Pa s m-1]
   real :: b_denom_1   ! The first term in the denominator of b1 [H ~> m or kg m-2].
+  real, dimension(SZIB_(G),SZJ_(G),SZK_(GV)) &
+                          :: visc_rem_u_local !< Fraction of a time-step's worth of a
```
this was bizarre, trying to use visc_rem_u and visc_rem_v directly from the variables that came into the function would lead to failures and bizarre behaviour at runtime. Creating a local copy and allocating it fixed it. Maybe they are on the GPU from somewhere and I need to use an `update` clause?

marshallward replied:
Once things are stable, we should investigate this further before merging.  Often these are a sign of out-of-sync data.

JorgeG94 replied:
indeed, 100% what I found today is the reason behind my data needing to be either mapped explicitly or me having to create temp arrays is because in this function and in others, a derived type is being passed in as:
```
subroutine vertvisc_remnant(visc, visc_rem_u, visc_rem_v, dt, G, GV, US, CS)
```

and called as 

```
call vertvisc_remnant(visc, CS%visc_rem_u, CS%visc_rem_v, dt, G, GV, US, CS%vertvisc_CSp)
``` 

So the variables that are giving me grief are not direct references to visc_rem_u/v or CS they are arrays inside the derived type being passed in OR a derived type inside a derived type so I think openmp is getting confuso

edoyango replied:
I haven't encountered any issues with `CS%visc_rem_u/v` when working on the continuity solver (which also passes in those arrays in the same manner). Although they're `intent(in)` for continuity and `intent(inout)` for visc_remnant...

JorgeG94 commented on src/parameterizations/vertical/MOM_vert_friction.F90:
```diff
@@ -1285,10 +1604,16 @@ subroutine vertvisc_remnant(visc, visc_rem_u, visc_rem_v, dt, G, GV, US, CS)
+  integer :: jsc, jec
   is = G%isc ; ie = G%iec
   Isq = G%IscB ; Ieq = G%IecB ; Jsq = G%JscB ; Jeq = G%JecB ; nz = GV%ke
+  jsc = G%jsc ; jec = G%jec
```
this seemed convenient and in alignment with the other ones created here

marshallward replied:
:+1: Should be ok, although note that it could complicate rebasing with dev/gfdl.  

If you do want to keep this:

* Use `js` and `je`, following `is` and `ie`.
* Integrate declaration in L1613 with L1612

My suggestion is to defer these style improvements, if only because this is a rather epochal refactor and fewer changes makes it easier to review and verify the index bounds.

JorgeG94 commented on src/parameterizations/vertical/MOM_vert_friction.F90:
```diff
@@ -1297,43 +1622,57 @@ subroutine vertvisc_remnant(visc, visc_rem_u, visc_rem_v, dt, G, GV, US, CS)
 
-  do j=G%jsc,G%jec ; do I=Isq,Ieq ; if (G%mask2dCu(I,j) > 0.) then
+  !! TODO JORGE: PORT
+  !$omp target teams loop collapse(2) map(to: CS%h_u, CS%a_u, G, G%mask2dCu) 
```
In find_coupling_coef I had some nightmares with derived types arrays being an insane pain unless I mapped them explicitly in the loop. _no idea why_ this should disappear with do concurrent, I believe

marshallward replied:
Before merging, we should try to figure out a way to eliminate these once things are working.

JorgeG94 commented on src/parameterizations/vertical/MOM_vert_friction.F90:
```diff
@@ -1361,21 +1704,33 @@ subroutine vertvisc_remnant(visc, visc_rem_u, visc_rem_v, dt, G, GV, US, CS)
+  !$omp target update from(b1, c1, d1, Ray, visc_rem_u_local, visc_rem_v_local)
+  !$omp target exit data map(delete: b1, c1, d1, Ray, visc_rem_u_local, visc_rem_v_local)
+  !JORGE TODO: figure out why I cannot use the intent inout variable
+  visc_rem_u = visc_rem_u_local
```
this could be a loopy loop loop for _performance_

marshallward replied:
The MOM6 style guide requires indices for array copies. 

This should be OK: `visc_rem_u(:,:,:) = visc_rem_u_local(:,:,:)`

(Hopefully, we can eliminate the introduction of `visc_rem_u_local`, especially since we already have an excess of 3d arrays.)

JorgeG94 commented on src/parameterizations/vertical/MOM_vert_friction.F90:
```diff
@@ -1440,6 +1795,7 @@ subroutine vertvisc_coef(u, v, h, dz, forces, visc, tv, dt, G, GV, US, CS, OBC,
   real, dimension(SZIB_(G),SZJB_(G)) :: &
     kv_bbl, &     ! The bottom boundary layer viscosity [H Z T-1 ~> m2 s-1 or Pa s].
     bbl_thick, &  ! The bottom boundary layer thickness [Z ~> m].
+    test_bbl_thick, &  ! The bottom boundary layer thickness [Z ~> m].
```
delete

JorgeG94 commented on src/parameterizations/vertical/MOM_vert_friction.F90:
```diff
@@ -1551,17 +1907,41 @@ subroutine vertvisc_coef(u, v, h, dz, forces, visc, tv, dt, G, GV, US, CS, OBC,
+  endif
+
+  !$omp target enter data map(to: G, G%bathyT)
+  !!$omp target enter data map(to: visc, visc%bbl_thick_u)
```
reference from comment above on certain derived types giving me nightmares. If commented back in, answers would oscillate and the bathyT  variable would cause problems in the simulation

JorgeG94 commented on src/parameterizations/vertical/MOM_vert_friction.F90:
```diff
@@ -1590,38 +1970,62 @@ subroutine vertvisc_coef(u, v, h, dz, forces, visc, tv, dt, G, GV, US, CS, OBC,
+        end do 
+      end do 
+      !$omp end target teams loop
+    end do 
```
this is some defensive array initialization because I am paranoid - probably not needed!!!!

JorgeG94 commented on src/parameterizations/vertical/MOM_vert_friction.F90:
```diff
@@ -1590,38 +1970,62 @@ subroutine vertvisc_coef(u, v, h, dz, forces, visc, tv, dt, G, GV, US, CS, OBC,
+      end do 
+      !$omp end target teams loop
+    end do 
+  do k=nz,1,-1
```
need to check if I can move the k loop inside the target region, will I die?

marshallward replied:
Despite the performance implications, I think we should leave k outside for now.  The k-loop is almost acting as an implicit method to barrier the kernels from each other, and a user will typically skip most of them.

JorgeG94 commented on src/parameterizations/vertical/MOM_vert_friction.F90:
```diff
@@ -1719,9 +2132,17 @@ subroutine vertvisc_coef(u, v, h, dz, forces, visc, tv, dt, G, GV, US, CS, OBC,
+
+
+  ! enters z_i, dz_vel, 
+  ! ported, works
```
delete

JorgeG94 commented on src/parameterizations/vertical/MOM_vert_friction.F90:
```diff
@@ -1734,41 +2155,41 @@ subroutine vertvisc_coef(u, v, h, dz, forces, visc, tv, dt, G, GV, US, CS, OBC,
 
   do_any_shelf = .false.
-  if (associated(forces%frac_shelf_u)) then
+  if (associated(forces%frac_shelf_u)) then !{
```
ah yes, this way I can cntrl % with vim around this gigantic if statements, probably delete

JorgeG94 commented on src/parameterizations/vertical/MOM_vert_friction.F90:
```diff
@@ -1893,20 +2313,24 @@ subroutine vertvisc_coef(u, v, h, dz, forces, visc, tv, dt, G, GV, US, CS, OBC,
+    endif !}
 
     do K=1,nz+1
+    !$omp target teams loop collapse(2) map(tofrom: CS, CS%a_u)
```
fix mapping issue

JorgeG94 commented on src/parameterizations/vertical/MOM_vert_friction.F90:
```diff
@@ -1969,36 +2405,59 @@ subroutine vertvisc_coef(u, v, h, dz, forces, visc, tv, dt, G, GV, US, CS, OBC,
+  !$omp target update from(Dmin, zi_dir)
+  !$omp target update from(dz_harm) 
+  block 
+  real :: tmp1, tmp2
```
probably remove, I did this because I thought I was having issues while mapping the variables tmp are using down in this loop

JorgeG94 commented on src/parameterizations/vertical/MOM_vert_friction.F90:
```diff
@@ -1969,36 +2405,59 @@ subroutine vertvisc_coef(u, v, h, dz, forces, visc, tv, dt, G, GV, US, CS, OBC,
+
   do k=nz,1,-1
+
+    !!$omp target teams loop collapse(2) firstprivate(k) 
```
oh yeah this is the dead loops of hell, they have not been ported!!! attention!

JorgeG94 commented on src/parameterizations/vertical/MOM_vert_friction.F90:
```diff
@@ -2673,10 +3194,14 @@ subroutine find_coupling_coef(a_cpl, hvel, do_i, h_harm, bbl_thick, kv_bbl, z_i,
+  ! JORGE TODO: PORT TO GPU
 
     ! Find the friction velocity and the absolute value of the Coriolis parameter at this point.
+    !$omp target
     u_star(:,:) = 0.0  ! Zero out the friction velocity on land points.
+    absf(:,:) = 0.0
     tau_mag(:,:) = 0.0  ! Zero out the friction velocity on land points.
+    !$omp end target
```
paranoid target region to inializae the array to 0, probably can be done better

edoyango commented on src/parameterizations/vertical/MOM_vert_friction.F90:
```diff
@@ -774,414 +817,689 @@ subroutine vertvisc(u, v, h, forces, visc, dt, OBC, ADp, CDp, G, GV, US, CS, &
   !$omp target enter data map(to: CS, CS%a_u, CS%h_u)
-
   !$omp target enter data map(alloc: b1, c1, d1)
+block 
```
blocks aren't used anywhere else in MOM6, so you'll probably want to move the variable declarations to the top. Also helps with rebasing.

edoyango commented on src/parameterizations/vertical/MOM_vert_friction.F90:
```diff
@@ -774,414 +817,689 @@ subroutine vertvisc(u, v, h, forces, visc, dt, OBC, ADp, CDp, G, GV, US, CS, &
+      !$omp end target teams loop
       enddo
 
-      ! back substitute to solve for new velocities, held by ADp%du_dt_visc_gl90
```
Want to make sure the comments are preserved.

edoyango commented on src/parameterizations/vertical/MOM_vert_friction.F90:
```diff
@@ -689,50 +707,75 @@ subroutine vertvisc(u, v, h, forces, visc, dt, OBC, ADp, CDp, G, GV, US, CS, &
+  !$omp target enter data map(alloc: ADp%du_dt_visc_gl90)
+
+  if (associated(ADp%du_dt_visc_gl90)) then
+    do k=1,nz
```
While I very much prefer the staggered formatting of loops rather than the inline thing MOM6 has, it would likely be much easier to rebase against dev/gfdl if you kept the original loop format. Although might be moot if you convert to do concurrent.

edoyango commented on src/parameterizations/vertical/MOM_vert_friction.F90:
```diff
@@ -2026,31 +2485,38 @@ subroutine vertvisc_coef(u, v, h, dz, forces, visc, tv, dt, G, GV, US, CS, OBC,
       ! Montgomery potential gradients at the bottom where nearly massless
-      ! layers ride over the topography.
 
+      !!$omp target teams loop collapse(2) private(i,j, tmp1, tmp2, z2, botfn) firstprivate(k)
```
Was this the problem loop?

edoyango commented on src/parameterizations/vertical/MOM_vert_friction.F90:
```diff
@@ -1551,17 +1907,41 @@ subroutine vertvisc_coef(u, v, h, dz, forces, visc, tv, dt, G, GV, US, CS, OBC,
+  !!$omp target enter data map(to: visc, visc%bbl_thick_u)
+  !!$omp target update to(visc, visc%bbl_thick_u)
+  !!$omp target enter data map(to: CS, CS%harm_BL_val)
+  !$omp target update to(u,h,dz)
```
FYI this is needed because there is `update from()` in `step_MOM_dyn_split_RK2` (i.e., the main loop).

marshallward replied:
`update from ()` should not change the values, even if it's redundant.  Do you mean there is additiona CPU work after the `update from ()` in the dycore loop?

Regardless, we should be able to establish the status of `u`, `h`, `dz` outside of the function.

edoyango replied:
sorry I wasn't clear - my comment was meant as a hint for removing data transfers in the dyncore loop. `u,h,dz` are `update from` to make sure vert friction routines work correctly on the CPU.

marshallward commented on src/parameterizations/vertical/MOM_vert_friction.F90:
```diff
@@ -1,7 +1,24 @@
+  call nvtxEndRange
+end subroutine end_nvtx
+
+end module mom_nvtx
```
Reminder to remove the nvtx profiler content when it's ready for merge.  (Mostly a note to myself.,)

marshallward commented on src/parameterizations/vertical/MOM_vert_friction.F90:
```diff
@@ -689,50 +707,75 @@ subroutine vertvisc(u, v, h, forces, visc, dt, OBC, ADp, CDp, G, GV, US, CS, &
+  ! this could be a function 
   !$omp target enter data map(to: ADp)
+  !$omp target enter data map(to: G, G%mask2dCu)
+  !$omp target update to(u)
```
Ideally, `u` should be set up outside of the function, including updates.

Similarly, `G` and its subarrays should be "global" and available for the entire run, so it should be OK to drop.

JorgeG94 replied:
yeah a lot of this is me just trying to get things to work !

edoyango replied:
the `target update to(u)` is necessary atm since there's loops in the dyncore loop that update `u,v` on the CPU before entering vert friction routines. Should be able to remove the `update to(u)` once the dyncore loop is ported.

marshallward commented on src/parameterizations/vertical/MOM_vert_friction.F90:
```diff
@@ -689,50 +707,75 @@ subroutine vertvisc(u, v, h, forces, visc, dt, OBC, ADp, CDp, G, GV, US, CS, &
+        end if
+      end do
+    end do
+    !$omp end target teams loop
```
I'd suggest reverting to do-concurrent if it works correctly (though perhaps that is your plan.)

Also, I'd the TODO comment, or perhaps edit it, to reflect that `ADp%du_dt_str` should ideally be separated from the loop.  (Perhaps the k-loop could be moved outside?  I know we're moving away from that direction.  This may need more discussion...)

Following @edoyango's comment above, try to preserve the existing formatting to the best of your ability, particularly semicolons and whitespace.

JorgeG94 replied:
yeah, DO concurrent is the plan this was mostly for debugging since I needed to pinpoint where was my data getting destroyed and turning off do concurrent is a bit more difficult than openmp

marshallward commented on src/parameterizations/vertical/MOM_vert_friction.F90:
```diff
@@ -2950,9 +3504,19 @@ subroutine find_coupling_coef(a_cpl, hvel, do_i, h_harm, bbl_thick, kv_bbl, z_i,
+  !$omp target exit data map(from: a_cpl, h_ml, absf, nk_in_ml)
+  !$omp target exit data map(delete: u_star, tau_mag)
+  !$omp target exit data map(delete: absf)
+  !$omp target exit data map(delete: z_t, Kv_tot, Kv_add)
```
As above, ideally we want these outputs to be set up outside of the functions.  Individual functions should not need to do any data transfers.  For now, this is delegated to the dycore.  (Eventually, even the dycore should have none.)

marshallward commented on src/parameterizations/vertical/MOM_vert_friction.F90:
```diff
@@ -3070,6 +3639,7 @@ subroutine vertvisc_limit_vel(u, v, h, ADp, CDp, forces, visc, dt, G, GV, US, CS
   endif
 
   if (len_trim(CS%u_trunc_file) > 0) then
+    ! JORGE TODO: PORT?
```
These only appear if there is a problem, usually a numerical instability.  I would skip for now (but perhaps do a !$omp target update to() if any CPU fields are needed).

(comment applies to all CFL truncation blocks)

marshallward reviewed (changes requested):
@JorgeG94 I did a first pass of the code.  Some initial thoughts to help streamline the review:

* Restore the original style where possible (esp whitespace and semicolons).

* Complete the `!$omp target` -> `do concurrent` conversions where necessary

* Trim out any redundant debugging code or comments, including `ntfx` content. 

Assuming everything else works, that would help to review some of the more complicated issues that have arised.

Technical issues:

* Eliminate blocks.  If they are needed for numerical reasons, we should start with 

* Restore CPU-side `if (associated(ADp%du_dt_str))` etc checks.  If they don't work for a particular compiler version, then we may want to avoid that version until it's fixed.  As mentioned in the inline comments, this is a widely used pattern.

* Can we eliminate `visc_rem_[uv]_local`?  This feels like fairly high priority.

Long term:

* Minimize `map(to:), `update()` inside of the individual functions.  (Eventually this should be zero, as in the other functions.)

* Separate conditional content from do-concurrent loops. `ADp%d[uv]_dt_str` is the one that comes to mind, but there may be others.

* Remove OpenMP/do concurrent loops for untested blocks.  The OBCs come to mind.  Such blocks should probably also have `!$omp target update to/from` (although these won't be tested either!)

  (Also, I may have a small ice shelf test case to run when we're ready.  But I agree that we should not try to port any of this code yet.)

No need to do all of this at once.  And let me know if you'd like me to send a patch to your PR with any of these changes.

JorgeG94 committed:
back to format and delete block constructs

JorgeG94 commented on src/parameterizations/vertical/MOM_vert_friction.F90:
```diff
@@ -774,402 +811,554 @@ subroutine vertvisc(u, v, h, forces, visc, dt, OBC, ADp, CDp, G, GV, US, CS, &
       enddo
     endif
 
     do concurrent (j=G%isc:G%jec, I=Isq:Ieq, G%mask2dCu(I,j) > 0.)
```
@marshallward I just noticed that this loop goes from `isc to jec` is this correct? or should it be jsc:jec?

marshallward replied:
Yikes!  Yes that should be `j=G%jsc,G%jec` (or the equivalent `do concurrent`).

I see it in `dev/gpu` but not in `dev/gfdl`, so I must have introduced it.  It may have slipped through testing because we are using local internal indexing for each domain (G%isc = G%jsc = 5).  Either way, still a bad error!

Thanks for catching that!

JorgeG94 committed:
remove the maps ins remnant yay!

JorgeG94 committed:
remove need for local vertvisc_u/v variables

JorgeG94 committed:
death loop of death has been ported

JorgeG94 committed:
vertvisc coef does not have explicit mappings anymore

JorgeG94 committed:
the maps are gone!

JorgeG94 committed:
limit vel using single data transfer

JorgeG94 committed:
do concurrent

JorgeG94 committed:
some memory cleanup and code cleanup - profiling time

JorgeG94 committed:
update and lots of nvtx markers for opt

JorgeG94 committed:
limit vel betterments and notes for optimization

JorgeG94 committed:
make limit vel good

JorgeG94 commented on src/parameterizations/vertical/MOM_vert_friction.F90:
```diff
@@ -2026,32 +2224,32 @@ subroutine vertvisc_coef(u, v, h, dz, forces, visc, tv, dt, G, GV, US, CS, OBC,
         dz_vel(i,J,k) = dz_harm(i,J,k)
 
-        if (v(i,J,k) * h_delta(i,J) < 0) then
+        tmp1 = v(i,j,k)
```
I probably don't need this

JorgeG94 committed:
h_ml transfer

marshallward committed:
Vertfrict: Move CS allocations to init

Allocations of the vert frictions and visc control structures were moved
from the dycore subroutine loops to the model initialization.

Redundant grid (G) transfers were removed (although they were not
triggering any transfers).

Most (but not all) visc arrays have also been moved into the
initialization.  Kv_shear and Ray_[uv] need to be added.

Trailing whitespace also (inadvertently) got removed, hopefully not a
distraction.

edoyango commented on src/parameterizations/vertical/MOM_vert_friction.F90:
```diff
@@ -2992,62 +3279,88 @@ subroutine vertvisc_limit_vel(u, v, h, ADp, CDp, forces, visc, dt, G, GV, US, CS
 
-      if (trunc_any) then ; if (CS%CFL_based_trunc) then
-        do k=1,nz ; do I=Isq,Ieq
+      do concurrent (j=js:je, k=1:nz, i=isq:ieq, trunc_any_array(i,j,k))
```
I can see that you made trunc_any 3D, but you can probably just get rid of it altogether since (from what I can tell), trunc_any was just used to stop the block from running unnecessarily.

JorgeG94 replied:
I am not sure about this O.o need to look at the code better

JorgeG94 committed:
Merge branch 'dev/gpu' into jorge/dev_gpu_vertvisc

JorgeG94 committed:
Merge branch 'jorge/dev_gpu_vertvisc' into vert_frict_cs_alloc

JorgeG94 committed:
Merge pull request #1 from marshallward/vert_frict_cs_alloc

Vertfrict: Move CS allocations to init

marshallward committed:
coupling coeff fixes

This appears to improve the CPU/GPU repro of the latest merge of dev/gpu
into dev_gpu_vertvisc.

* Resolves some of the u/v/h/dz field states between the barotropic and
  vertvisc functions.  (Development of each assumed that the other was
  not complete).

* This seems to fix some issues with a_[uv], Kv_tot, and dz inside of
  find_coupling_coef and vertvisc coef

  * chksum transfers were added to ensure consistent

* The diff has moved to frhat[uv] in btcalc.  The fields are all zero on
  GPU, so they are probably not transferred.

Work in progress...

marshallward committed:
Remove redundant BT_cont transfer

The BT_cont fields are now on GPU, so they don't need to be transferred
before the btcalc call.

Checksum transfers for h_ml were also added.  Still assuming that it's
computed on the CPU.

JorgeG94 committed:
Merge pull request #2 from marshallward/dev_gpu_vertvisc_merge_fix

hell yeah

JorgeG94 committed:
delete nvtcx

JorgeG94 committed:
heavy memory copies but loops in rk loop are on the gpu

JorgeG94 committed:
cleanup some comments

JorgeG94 committed:
restore gpu enabling for thickness to dz

JorgeG94 committed:
test to see if guarding and reintroducing the check helps

JorgeG94 committed:
an ugly fix for a simple problem: just want to be sure before I dive

JorgeG94 committed:
spaces

JorgeG94 committed:
Merge branch 'jorge/dev_gpu_vertvisc' into jorge/main_rk_loop

JorgeG94 committed:
formatting

JorgeG94 committed:
target

JorgeG94 commented on src/parameterizations/vertical/MOM_vert_friction.F90:
```diff
@@ -1579,13 +1733,13 @@ subroutine vertvisc_coef(u, v, h, dz, forces, visc, tv, dt, G, GV, US, CS, OBC,
   do k=nz,1,-1
-    do j=js,je ; do I=Isq,Ieq ; if (do_i(I,j)) then
+    do concurrent (j=js:je, i=isq:ieq, do_i(i,j))
       h_harm(I,j) = 2. * h(i,j,k) * h(i+1,j,k) / (h(i,j,k) + h(i+1,j,k) + h_neglect)
```
@marshallward this iteration loops go over I but the variables inside use i, is this wrong?

marshallward replied:
It's correct.  h is center-indexed and is being interpolated to h_harm which is face-indexed.  The loop is a face-loop so I is uppercase.

This sign convention is hard to get right.  Fow now I'd suggest just mirroring the target branch as best as you can.  We can go over it at the end.

JorgeG94 replied:
excellent, I'll finish this tomorrow

JorgeG94 committed:
isq -> Isq

JorgeG94 committed:
jsq -> jSQ

JorgeG94 committed:
remove curly brances

JorgeG94 committed:
delete tmp

JorgeG94 committed:
one more jsq that was still small

JorgeG94 committed:
fix nvhpc cpu ci

JorgeG94 committed:
more Jsq and Isq

JorgeG94 committed:
Merge pull request #3 from JorgeG94/jorge/main_rk_loop

heavy memory copies but loops in rk loop are on the gpu

marshallward commented:
Handled in #44
